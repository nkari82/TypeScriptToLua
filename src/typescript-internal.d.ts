export {};

declare module "typescript" {
    function createDiagnosticReporter(system: System, pretty?: boolean): DiagnosticReporter;
    function createWatchStatusReporter(system: System, pretty?: boolean): WatchStatusReporter;

    interface System {
        setBlocking?(): void;
    }

    interface Statement {
        jsDoc?: JSDoc[];
    }

    interface Program {
        getCommonSourceDirectory(): string;
    }

    interface CompilerOptions {
        configFile?: TsConfigSourceFile;
        configFilePath?: string;
    }

    interface TypeChecker {
        getElementTypeOfArrayType(type: Type): Type | undefined;
        getContextualTypeForObjectLiteralElement(element: ObjectLiteralElementLike): Type | undefined;

        isTupleType(type: Type): boolean;
        isArrayType(type: Type): boolean;
    }

    interface Symbol {
        // false positive
        // eslint-disable-next-line @typescript-eslint/ban-types
        parent?: Symbol;
    }

    function transformJsx(context: TransformationContext): (x: SourceFile) => SourceFile;

    export type OuterExpression =
        | ParenthesizedExpression
        | TypeAssertion
        | AsExpression
        | NonNullExpression
        | PartiallyEmittedExpression;

    function skipOuterExpressions(node: Expression, kinds?: OuterExpressionKinds): Expression;
    export function isOuterExpression(node: Node, kinds?: OuterExpressionKinds): node is OuterExpression;
}
