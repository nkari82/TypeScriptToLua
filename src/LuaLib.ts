import * as path from "path";
import { EmitHost } from "./transpilation";
import * as lua from "./LuaAST";

export enum LuaLibFeature {
    ArrayConcat = "ArrayConcat",
    ArrayEntries = "ArrayEntries",
    ArrayEvery = "ArrayEvery",
    ArrayFilter = "ArrayFilter",
    ArrayForEach = "ArrayForEach",
    ArrayFind = "ArrayFind",
    ArrayFindIndex = "ArrayFindIndex",
    ArrayFrom = "ArrayFrom",
    ArrayIncludes = "ArrayIncludes",
    ArrayIndexOf = "ArrayIndexOf",
    ArrayIsArray = "ArrayIsArray",
    ArrayJoin = "ArrayJoin",
    ArrayMap = "ArrayMap",
    ArrayPush = "ArrayPush",
    ArrayPushArray = "ArrayPushArray",
    ArrayReduce = "ArrayReduce",
    ArrayReduceRight = "ArrayReduceRight",
    ArrayReverse = "ArrayReverse",
    ArrayUnshift = "ArrayUnshift",
    ArraySort = "ArraySort",
    ArraySlice = "ArraySlice",
    ArraySome = "ArraySome",
    ArraySplice = "ArraySplice",
    ArrayToObject = "ArrayToObject",
    ArrayFlat = "ArrayFlat",
    ArrayFlatMap = "ArrayFlatMap",
    ArraySetLength = "ArraySetLength",
    Await = "Await",
    Class = "Class",
    ClassExtends = "ClassExtends",
    CloneDescriptor = "CloneDescriptor",
    Decorate = "Decorate",
    DecorateParam = "DecorateParam",
    Delete = "Delete",
    DelegatedYield = "DelegatedYield",
    Error = "Error",
    FunctionBind = "FunctionBind",
    Generator = "Generator",
    InstanceOf = "InstanceOf",
    InstanceOfObject = "InstanceOfObject",
    Iterator = "Iterator",
    LuaIteratorSpread = "LuaIteratorSpread",
    Map = "Map",
    MathAtan2 = "MathAtan2",
    MathSign = "MathSign",
    New = "New",
    Number = "Number",
    NumberIsFinite = "NumberIsFinite",
    NumberIsNaN = "NumberIsNaN",
    NumberToString = "NumberToString",
    ObjectAssign = "ObjectAssign",
    ObjectDefineProperty = "ObjectDefineProperty",
    ObjectEntries = "ObjectEntries",
    ObjectFromEntries = "ObjectFromEntries",
    ObjectGetOwnPropertyDescriptor = "ObjectGetOwnPropertyDescriptor",
    ObjectGetOwnPropertyDescriptors = "ObjectGetOwnPropertyDescriptors",
    ObjectKeys = "ObjectKeys",
    ObjectRest = "ObjectRest",
    ObjectValues = "ObjectValues",
    ParseFloat = "ParseFloat",
    ParseInt = "ParseInt",
    Promise = "Promise",
    PromiseAll = "PromiseAll",
    PromiseAllSettled = "PromiseAllSettled",
    PromiseAny = "PromiseAny",
    PromiseRace = "PromiseRace",
    Set = "Set",
    SetDescriptor = "SetDescriptor",
    SparseArrayNew = "SparseArrayNew",
    SparseArrayPush = "SparseArrayPush",
    SparseArraySpread = "SparseArraySpread",
    WeakMap = "WeakMap",
    WeakSet = "WeakSet",
    SourceMapTraceBack = "SourceMapTraceBack",
    Spread = "Spread",
    StringAccess = "StringAccess",
    StringCharAt = "StringCharAt",
    StringCharCodeAt = "StringCharCodeAt",
    StringEndsWith = "StringEndsWith",
    StringIncludes = "StringIncludes",
    StringPadEnd = "StringPadEnd",
    StringPadStart = "StringPadStart",
    StringReplace = "StringReplace",
    StringReplaceAll = "StringReplaceAll",
    StringSlice = "StringSlice",
    StringSplit = "StringSplit",
    StringStartsWith = "StringStartsWith",
    StringSubstr = "StringSubstr",
    StringSubstring = "StringSubstring",
    StringTrim = "StringTrim",
    StringTrimEnd = "StringTrimEnd",
    StringTrimStart = "StringTrimStart",
    Symbol = "Symbol",
    SymbolRegistry = "SymbolRegistry",
    TypeOf = "TypeOf",
    Unpack = "Unpack",
}

export interface LuaLibFeatureInfo {
    dependencies?: LuaLibFeature[];
    exports: string[];
}
export type LuaLibModulesInfo = Record<LuaLibFeature, LuaLibFeatureInfo>;

export const luaLibModulesInfoFileName = "lualib_module_info.json";
let luaLibModulesInfo: LuaLibModulesInfo | undefined;
export function getLuaLibModulesInfo(emitHost: EmitHost): LuaLibModulesInfo {
    if (luaLibModulesInfo === undefined) {
        const lualibPath = path.resolve(__dirname, `../dist/lualib/${luaLibModulesInfoFileName}`);
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibModulesInfo = JSON.parse(result) as LuaLibModulesInfo;
        } else {
            throw new Error(`Could not load lualib dependencies from '${lualibPath}'`);
        }
    }
    return luaLibModulesInfo;
}

export function readLuaLibFeature(feature: LuaLibFeature, emitHost: EmitHost): string {
    const featurePath = path.resolve(__dirname, `../dist/lualib/${feature}.lua`);
    const luaLibFeature = emitHost.readFile(featurePath);
    if (luaLibFeature === undefined) {
        throw new Error(`Could not load lualib feature from '${featurePath}'`);
    }
    return luaLibFeature;
}

export function resolveRecursiveLualibFeatures(
    features: Iterable<LuaLibFeature>,
    emitHost: EmitHost,
    luaLibModulesInfo: LuaLibModulesInfo = getLuaLibModulesInfo(emitHost)
): LuaLibFeature[] {
    const loadedFeatures = new Set<LuaLibFeature>();
    const result: LuaLibFeature[] = [];

    function load(feature: LuaLibFeature): void {
        if (loadedFeatures.has(feature)) return;
        loadedFeatures.add(feature);

        const dependencies = luaLibModulesInfo[feature]?.dependencies;
        if (dependencies) {
            dependencies.forEach(load);
        }

        result.push(feature);
    }

    for (const feature of features) {
        load(feature);
    }

    return result;
}

export function loadInlineLualibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): string {
    let result = "";

    for (const feature of resolveRecursiveLualibFeatures(features, emitHost)) {
        const luaLibFeature = readLuaLibFeature(feature, emitHost);
        result += luaLibFeature + "\n";
    }

    return result;
}

export function loadImportedLualibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): lua.Statement[] {
    const luaLibModuleInfo = getLuaLibModulesInfo(emitHost);

    const imports = Array.from(features).flatMap(feature => luaLibModuleInfo[feature].exports);

    const requireCall = lua.createCallExpression(lua.createIdentifier("require"), [
        lua.createStringLiteral("lualib_bundle"),
    ]);
    if (imports.length === 0) {
        return [];
    }

    const luaLibId = lua.createIdentifier("____lualib");
    const importStatement = lua.createVariableDeclarationStatement(luaLibId, requireCall);
    const statements: lua.Statement[] = [importStatement];
    // local <export> = ____luaLib.<export>
    for (const item of imports) {
        statements.push(
            lua.createVariableDeclarationStatement(
                lua.createIdentifier(item),
                lua.createTableIndexExpression(luaLibId, lua.createStringLiteral(item))
            )
        );
    }
    return statements;
}

let luaLibBundleContent: string;
export function getLuaLibBundle(emitHost: EmitHost): string {
    if (luaLibBundleContent === undefined) {
        const lualibPath = path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua");
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibBundleContent = result;
        } else {
            throw new Error(`Could not load lualib bundle from '${lualibPath}'`);
        }
    }

    return luaLibBundleContent;
}
