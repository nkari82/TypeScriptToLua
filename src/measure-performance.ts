import { performance } from "perf_hooks";

// We use our own performance hooks implementation for easier use, but also call node's performance hooks, so it shows up in the profiler.

let enabled = false;
const marks = new Map<string, number>();
const durations = new Map<string, number>();

function timestamp() {
    return performance.now();
}

/**
 * Marks a performance event, with the given markName.
 */
function mark(markName: string) {
    if (enabled) {
        marks.set(markName, timestamp());
        performance.mark(markName);
    }
}

/**
 * Adds a performance measurement with the specified name.
 *
 * @param measureName The name of the performance measurement.
 * @param startMarkName The name of the starting mark
 * @param endMarkName The name of the ending mark
 */
function measure(measureName: string, startMarkName: string, endMarkName: string) {
    if (enabled) {
        const end = marks.get(endMarkName) ?? timestamp();
        const start = marks.get(startMarkName) ?? performance.timeOrigin;
        const previousDuration = durations.get(measureName) ?? 0;
        durations.set(measureName, previousDuration + (end - start));
        performance.measure(measureName, startMarkName, endMarkName);
    }
}

/**
 * Starts a performance measurement section.
 * @param name name of the measurement
 */
export function startSection(name: string) {
    mark("start " + name);
}

/**
 * Ends a performance measurement section.
 * @param name name of the measurement
 */
export function endSection(name: string) {
    mark("end " + name);
    measure(name, "start " + name, "end " + name);
}

export function isMeasurementEnabled() {
    return enabled;
}

export function enableMeasurement() {
    if (!enabled) {
        enabled = true;
    }
}

export function disableMeasurement() {
    if (enabled) {
        enabled = false;
        marks.clear();
        durations.clear();
    }
}

export function forEachMeasure(callback: (measureName: string, duration: number) => void) {
    durations.forEach((duration, measureName) => callback(measureName, duration));
}

export function getTotalDuration() {
    let total = 0;
    forEachMeasure((_, duration) => (total += duration));
    return total;
}
