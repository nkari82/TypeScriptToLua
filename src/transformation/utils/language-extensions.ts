import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum ExtensionKind {
    MultiFunction = "MultiFunction",
    RangeFunction = "RangeFunction",
    VarargConstant = "VarargConstant",
    AdditionOperatorType = "Addition",
    AdditionOperatorMethodType = "AdditionMethod",
    SubtractionOperatorType = "Subtraction",
    SubtractionOperatorMethodType = "SubtractionMethod",
    MultiplicationOperatorType = "Multiplication",
    MultiplicationOperatorMethodType = "MultiplicationMethod",
    DivisionOperatorType = "Division",
    DivisionOperatorMethodType = "DivisionMethod",
    ModuloOperatorType = "Modulo",
    ModuloOperatorMethodType = "ModuloMethod",
    PowerOperatorType = "Power",
    PowerOperatorMethodType = "PowerMethod",
    FloorDivisionOperatorType = "FloorDivision",
    FloorDivisionOperatorMethodType = "FloorDivisionMethod",
    BitwiseAndOperatorType = "BitwiseAnd",
    BitwiseAndOperatorMethodType = "BitwiseAndMethod",
    BitwiseOrOperatorType = "BitwiseOr",
    BitwiseOrOperatorMethodType = "BitwiseOrMethod",
    BitwiseExclusiveOrOperatorType = "BitwiseExclusiveOr",
    BitwiseExclusiveOrOperatorMethodType = "BitwiseExclusiveOrMethod",
    BitwiseLeftShiftOperatorType = "BitwiseLeftShift",
    BitwiseLeftShiftOperatorMethodType = "BitwiseLeftShiftMethod",
    BitwiseRightShiftOperatorType = "BitwiseRightShift",
    BitwiseRightShiftOperatorMethodType = "BitwiseRightShiftMethod",
    ConcatOperatorType = "Concat",
    ConcatOperatorMethodType = "ConcatMethod",
    LessThanOperatorType = "LessThan",
    LessThanOperatorMethodType = "LessThanMethod",
    GreaterThanOperatorType = "GreaterThan",
    GreaterThanOperatorMethodType = "GreaterThanMethod",
    NegationOperatorType = "Negation",
    NegationOperatorMethodType = "NegationMethod",
    BitwiseNotOperatorType = "BitwiseNot",
    BitwiseNotOperatorMethodType = "BitwiseNotMethod",
    LengthOperatorType = "Length",
    LengthOperatorMethodType = "LengthMethod",
    TableNewType = "TableNew",
    TableDeleteType = "TableDelete",
    TableDeleteMethodType = "TableDeleteMethod",
    TableGetType = "TableGet",
    TableGetMethodType = "TableGetMethod",
    TableHasType = "TableHas",
    TableHasMethodType = "TableHasMethod",
    TableSetType = "TableSet",
    TableSetMethodType = "TableSetMethod",
    TableAddKeyType = "TableAddKey",
    TableAddKeyMethodType = "TableAddKeyMethod",
}
const extensionValues: Set<string> = new Set(Object.values(ExtensionKind));

export function getExtensionKindForType(context: TransformationContext, type: ts.Type): ExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlExtension");
    if (value && extensionValues.has(value)) {
        return value as ExtensionKind;
    }
}

const excludedTypeFlags: ts.TypeFlags =
    ((1 << 18) - 1) | // All flags from Any...Never
    ts.TypeFlags.Index |
    ts.TypeFlags.NonPrimitive;

function getPropertyValue(context: TransformationContext, type: ts.Type, propertyName: string): string | undefined {
    if (type.flags & excludedTypeFlags) return;
    const property = type.getProperty(propertyName);
    if (!property) return undefined;
    const propertyType = context.checker.getTypeOfSymbolAtLocation(property, context.sourceFile);
    if (propertyType.isStringLiteral()) return propertyType.value;
}

export function getExtensionKindForNode(context: TransformationContext, node: ts.Node): ExtensionKind | undefined {
    const originalNode = ts.getOriginalNode(node);
    let type = context.checker.getTypeAtLocation(originalNode);
    if (ts.isOptionalChain(originalNode)) {
        type = context.checker.getNonNullableType(type);
    }
    return getExtensionKindForType(context, type);
}

export function getExtensionKindForSymbol(
    context: TransformationContext,
    symbol: ts.Symbol
): ExtensionKind | undefined {
    const type = context.checker.getTypeOfSymbolAtLocation(symbol, context.sourceFile);
    return getExtensionKindForType(context, type);
}

export enum IterableExtensionKind {
    Iterable = "Iterable",
    Pairs = "Pairs",
    PairsKey = "PairsKey",
}

export function getIterableExtensionTypeForType(
    context: TransformationContext,
    type: ts.Type
): IterableExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlIterable");
    if (value && value in IterableExtensionKind) {
        return value as IterableExtensionKind;
    }
}

export function getIterableExtensionKindForNode(
    context: TransformationContext,
    node: ts.Node
): IterableExtensionKind | undefined {
    const type = context.checker.getTypeAtLocation(node);
    return getIterableExtensionTypeForType(context, type);
}
