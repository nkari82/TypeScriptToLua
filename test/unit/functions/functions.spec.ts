import * as ts from "typescript";
import * as tstl from "../../../src";
import * as util from "../../util";
import { unsupportedForTarget } from "../../../src/transformation/utils/diagnostics";

test("Arrow Function Expression", () => {
    util.testFunction`
        const add = (a, b) => a + b;
        return add(1, 2);
    `.expectToMatchJsResult();
});

test("Returning arrow function from arrow function", () => {
    util.testFunction`
        const add = (x: number) => (y: number) => x + y;
        return add(1)(2);
    `.expectToMatchJsResult();
});

test.each(["i++", "i--", "++i", "--i"])("Arrow function unary expression (%p)", lambda => {
    util.testFunction`
        let i = 10;
        [1,2,3,4,5].forEach(() => ${lambda});
        return i;
    `.expectToMatchJsResult();
});

test.each(["b => a = b", "b => a += b", "b => a -= b", "b => a *= b", "b => a /= b", "b => a **= b", "b => a %= b"])(
    "Arrow function assignment (%p)",
    lambda => {
        util.testFunction`
            let a = 10;
            let lambda = ${lambda};
            lambda(5);
            return a;
        `.expectToMatchJsResult();
    }
);

test.each([{ args: [] }, { args: [1] }, { args: [1, 2] }])("Arrow default values (%p)", ({ args }) => {
    util.testFunction`
        const add = (a = 3, b = 4) => a + b;
        return add(${util.formatCode(...args)});
    `.expectToMatchJsResult();
});

test("Function Expression", () => {
    util.testFunction`
        let add = function(a, b) {return a+b};
        return add(1,2);
    `.expectToMatchJsResult();
});

test("Function definition scope", () => {
    util.testFunction`
        function abc() { function xyz() { return 5; } }
        function def() { function xyz() { return 3; } abc(); return xyz(); }
        return def();
    `.expectToMatchJsResult();
});

test("Function default parameter", () => {
    util.testFunction`
        function abc(defaultParam: string = "abc") { return defaultParam; }
        return abc() + abc("def");
    `.expectToMatchJsResult();
});

test.each([{ inp: [] }, { inp: [5] }, { inp: [1, 2] }])("Function Default Values (%p)", ({ inp }) => {
    const callArgs = inp.join(",");

    util.testFunction(
        `let add = function(a: number = 3, b: number = 4) { return a+b; };
        return add(${callArgs});`
    ).expectToMatchJsResult();
});

test("Function default array binding parameter", () => {
    util.testFunction`
        function foo([bar]: [string] = ["foobar"]) {
            return bar;
        }
        return foo();
    `.expectToMatchJsResult();
});

test("Function default object binding parameter", () => {
    util.testFunction`
        function foo({ bar }: { bar: string } = { bar: "foobar" }) {
            return bar;
        }
        return foo();
    `.expectToMatchJsResult();
});

test("Function default binding parameter maintains order", () => {
    util.testFunction`
        const resultsA = [{x: "foo"}, {x: "baz"}];
        const resultsB = ["blah", "bar"];
        let i = 0;
        function a() { return resultsA[i++]; }
        function b() { return resultsB[i++]; }
        function foo({ x }: { x: string } = a(), y = b()) {
            return x + y;
        }
        return foo();
    `.expectToMatchJsResult();
});

test("Class method call", () => {
    util.testFunction`
        class TestClass {
            public classMethod(): number { return 4; }
        }

        const classInstance = new TestClass();
        return classInstance.classMethod();
    `.expectToMatchJsResult();
});

test("Class dot method call void", () => {
    util.testFunction`
        class TestClass {
            public dotMethod: () => number = () => 4;
        }

        const classInstance = new TestClass();
        return classInstance.dotMethod();
    `.expectToMatchJsResult();
});

test("Class dot method call with parameter", () => {
    util.testFunction`
        class TestClass {
            public dotMethod: (x: number) => number = x => 3 * x;
        }

        const classInstance = new TestClass();
        return classInstance.dotMethod(4);
    `.expectToMatchJsResult();
});

test("Class static dot method", () => {
    util.testFunction`
        class TestClass {
            public static dotMethod: () => number = () => 4;
        }

        return TestClass.dotMethod();
    `.expectToMatchJsResult();
});

test("Class static dot method with parameter", () => {
    util.testFunction`
        class TestClass {
            public static dotMethod: (x: number) => number = x => 3 * x;
        }

        return TestClass.dotMethod(4);
    `.expectToMatchJsResult();
});

const functionTypeDeclarations = [
    ["arrow", ": (...args: any) => any"],
    ["call signature", ": { (...args: any): any; }"],
    ["generic", ": Function"],
    ["inferred", ""],
];

test.each(functionTypeDeclarations)("Function bind (%s)", (_, type) => {
    util.testFunction`
        const abc${type} = function (this: { a: number }, a: string, b: string) { return this.a + a + b; }
        return abc.bind({ a: 4 }, "b")("c");
    `.expectToMatchJsResult();
});

test.each(functionTypeDeclarations)("Function apply (%s)", (_, type) => {
    util.testFunction`
        const abc${type} = function (this: { a: number }, a: string) { return this.a + a; }
        return abc.apply({ a: 4 }, ["b"]);
    `.expectToMatchJsResult();
});

// Fix #1226: https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1226
test.each(functionTypeDeclarations)("function apply without arguments should not lead to exception (%s)", (_, type) => {
    util.testFunction`
        const f${type} = function (this: number) { return this + 3; }
        return f.apply(4);
    `.expectToMatchJsResult();
});

// Issue #1218: https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1218
test.each(["() => 4", "undefined"])("prototype call on nullable function (%p)", value => {
    util.testFunction`
        function call(f?: () => number) {
            return f?.apply(3);
        }
        return call(${value});
    `
        .setOptions({ strictNullChecks: true })
        .expectToMatchJsResult();
});

test.each(functionTypeDeclarations)("Function call (%s)", (_, type) => {
    util.testFunction`
        const abc${type} = function (this: { a: number }, a: string) { return this.a + a; }
        return abc.call({ a: 4 }, "b");
    `.expectToMatchJsResult();
});

test.each([
    "function fn() {}",
    "function fn(x, y, z) {}",
    "function fn(x, y, z, ...args) {}",
    "function fn(...args) {}",
    "function fn(this: void) {}",
    "function fn(this: void, x, y, z) {}",
    "function fnReference(x, y, z) {} const fn = fnReference;",
    "const wrap = (fn: (...args: any[]) => any) => (...args: any[]) => fn(...args); const fn = wrap((x, y, z) => {});",
])("function.length (%p)", declaration => {
    util.testFunction`
        ${declaration}
        return fn.length;
    `.expectToMatchJsResult();
});

test.each([tstl.LuaTarget.Lua50, tstl.LuaTarget.Lua51, tstl.LuaTarget.Universal])(
    "function.length unsupported (%p)",
    luaTarget => {
        util.testFunction`
            function fn() {}
            return fn.length;
        `
            .setOptions({ luaTarget })
            .expectDiagnosticsToMatchSnapshot([unsupportedForTarget.code]);
    }
);

test("Recursive function definition", () => {
    util.testFunction`
        function f() { return typeof f; };
        return f();
    `.expectToMatchJsResult();
});

test("Recursive function expression", () => {
    util.testFunction`
        let f = function() { return typeof f; };
        return f();
    `.expectToMatchJsResult();
});

test("Wrapped recursive function expression", () => {
    util.testFunction`
        function wrap<T>(fn: T) { return fn; }
        let f = wrap(function() { return typeof f; }); return f();
    `.expectToMatchJsResult();
});

test("Recursive arrow function", () => {
    util.testFunction`
        let f = () => typeof f;
        return f();
    `.expectToMatchJsResult();
});

test("Wrapped recursive arrow function", () => {
    util.testFunction`
        function wrap<T>(fn: T) { return fn; }
        let f = wrap(() => typeof f);
        return f();
    `.expectToMatchJsResult();
});

test("Object method declaration", () => {
    util.testFunction`
        let o = { v: 4, m(i: number): number { return this.v * i; } };
        return o.m(3);
    `.expectToMatchJsResult();
});

// https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1266
test("Object method declaration used with non-null operator (#1266)", () => {
    util.testFunction`
        let o = { v: 4, m(i: number): number { return this.v * i; } };
        return o.m!(3);
    `.expectToMatchJsResult();
});

test.each([
    { args: ["bar"], expected: "foobar" },
    { args: ["baz", "bar"], expected: "bazbar" },
])("Function overload (%p)", ({ args, expected }) => {
    util.testFunction`
        class O {
            prop = "foo";
            method(s: string): string;
            method(this: void, s1: string, s2: string): string;
            method(s1: string) {
                if (typeof this === "string") {
                    return this + s1;
                }
                return this.prop + s1;
            }
        };
        const o = new O();
        return o.method(${util.formatCode(...args)});
    `.expectToEqual(expected);
});

test("Nested Function", () => {
    util.testFunction`
        class C {
            private prop = "bar";
            public outer() {
                const o = {
                    prop: "foo",
                    innerFunc: function() { return this.prop; },
                    innerArrow: () => this.prop
                };
                return o.innerFunc() + o.innerArrow();
            }
        }
        let c = new C();
        return c.outer();
    `.expectToMatchJsResult();
});

test.each([
    { s1: "abc", s2: "abc" },
    { s1: "abc", s2: "def" },
])("Dot vs Colon method call (%p)", ({ s1, s2 }) => {
    util.testFunction`
        class MyClass {
            dotMethod(this: void, s: string) {
                return s;
            }
            colonMethod(s: string) {
                return s;
            }
        }
        const inst = new MyClass();
        return inst.dotMethod("${s1}") == inst.colonMethod("${s2}");
    `.expectToMatchJsResult();
});

test("Element access call", () => {
    util.testFunction`
        class C {
            prop = "bar";
            method(s: string) { return s + this.prop; }
        }
        const c = new C();
        return c['method']("foo");
    `.expectToMatchJsResult();
});

test("Element access call no args", () => {
    util.testFunction`
        class C {
            prop = "bar";
            method() { return this.prop; }
        }
        const c = new C();
        return c['method']();
    `.expectToMatchJsResult();
});

test("Complex element access call", () => {
    util.testFunction`
        class C {
            prop = "bar";
            method(s: string) { return s + this.prop; }
        }
        function getC() { return new C(); }
        return getC()['method']("foo");
    `.expectToMatchJsResult();
});

test("Complex element access call no args", () => {
    util.testFunction`
        class C {
            prop = "bar";
            method() { return this.prop; }
        }
        function getC() { return new C(); }
        return getC()['method']();
    `.expectToMatchJsResult();
});

test("Complex element access call statement", () => {
    util.testFunction`
        let foo: string;
        class C {
            prop = "bar";
            method(s: string) { foo = s + this.prop; }
        }
        function getC() { return new C(); }
        getC()['method']("foo");
        return foo;
    `.expectToMatchJsResult();
});

test("Function local overriding export", () => {
    util.testModule`
        export const foo = 5;
        function bar(foo: number) {
            return foo;
        }
        export const result = bar(7);
    `
        .setReturnExport("result")
        .expectToMatchJsResult();
});

test("Function using global as this", () => {
    // Value is provided with top-level return with ts-ignore, because modules are always strict.
    // TODO: Provide a different builder kind for such tests?
    util.testModule`
        (globalThis as any).foo = "foo";
        function bar(this: any) {
            return this.foo;
        }

        // @ts-ignore
        return bar();
    `.expectToEqual("foo");
});

test("Function rest binding pattern", () => {
    util.testFunction`
        function bar(foo: string, ...[bar, baz]: [string, string]) {
            return bar + baz + foo;
        }
        return bar("abc", "def", "xyz");
    `.expectToMatchJsResult();
});

test("Function rest parameter", () => {
    util.testFunction`
        function foo(a: unknown, ...b: string[]) {
            return b.join("");
        }
        return foo("A", "B", "C", "D");
    `.expectToMatchJsResult();
});

test("Function nested rest parameter", () => {
    util.testFunction`
        function foo(a: unknown, ...b: string[]) {
            function bar() {
                return b.join("");
            }
            return bar();
        }
        return foo("A", "B", "C", "D");
    `.expectToMatchJsResult();
});

test("Function nested rest spread", () => {
    util.testFunction`
        function foo(a: unknown, ...b: string[]) {
            function bar() {
                const c = [...b];
                return c.join("");
            }
            return bar();
        }
        return foo("A", "B", "C", "D");
    `.expectToMatchJsResult();
});

test("Function rest parameter (unreferenced)", () => {
    util.testFunction`
        function foo(a: unknown, ...b: string[]) {
            return "foobar";
        }
        return foo("A", "B", "C", "D");
    `
        .tap(builder => expect(builder.getMainLuaCodeChunk()).not.toMatch("{...}"))
        .expectToMatchJsResult();
});

test("Function rest parameter (referenced in property shorthand)", () => {
    util.testFunction`
        function foo(a: unknown, ...b: string[]) {
            const c = { b };
            return c.b.join("");
        }
        return foo("A", "B", "C", "D");
    `.expectToMatchJsResult();
});

test("named function expression reference", () => {
    util.testFunction`
        const y = function x() {
            return { x: typeof x, y: typeof y };
        };

        return y();
    `.expectToMatchJsResult();
});

test("missing declaration name", () => {
    util.testModule`
        function () {}
    `.expectDiagnosticsToMatchSnapshot([1003]);
});

test("top-level function declaration is global", () => {
    // Can't be tested with expectToMatchJsResult because in JS that would not be global
    util.testModule`
        import './a';
        export const result = foo();
    `
        .addExtraFile("a.ts", 'function foo() { return "foo" }')
        .expectToEqual({ result: "foo" });
});

// https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1325
test("call expression should not throw (#1325)", () => {
    util.testModule`
        function test<T>(iterator:Iterator<T>) {
            iterator.return?.();
        }
    `
        // Note: does not reproduce without strict=true
        .setOptions({ target: ts.ScriptTarget.ESNext, strict: true })
        .expectToHaveNoDiagnostics();
});
