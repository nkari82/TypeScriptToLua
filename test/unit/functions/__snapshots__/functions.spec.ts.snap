// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function default parameter with value "null" 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function foo(self, x)
        return x
    end
    return foo(nil)
end
return ____exports"
`;

exports[`Function default parameter with value "undefined" 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function foo(self, x)
        return x
    end
    return foo(nil)
end
return ____exports"
`;

exports[`function.length unsupported ("5.0"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function fn(self)
    end
    return debug.getinfo(fn).nparams - 1
end
return ____exports"
`;

exports[`function.length unsupported ("5.0"): diagnostics 1`] = `"main.ts(3,20): error TSTL: function.length is/are not supported for target Lua 5.0."`;

exports[`function.length unsupported ("5.1"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function fn(self)
    end
    return debug.getinfo(fn).nparams - 1
end
return ____exports"
`;

exports[`function.length unsupported ("5.1"): diagnostics 1`] = `"main.ts(3,20): error TSTL: function.length is/are not supported for target Lua 5.1."`;

exports[`function.length unsupported ("universal"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local function fn(self)
    end
    return debug.getinfo(fn).nparams - 1
end
return ____exports"
`;

exports[`function.length unsupported ("universal"): diagnostics 1`] = `"main.ts(3,20): error TSTL: function.length is/are not supported for target Lua universal."`;

exports[`missing declaration name: code 1`] = `
"function ____(self)
end"
`;

exports[`missing declaration name: diagnostics 1`] = `"main.ts(2,18): error TS1003: Identifier expected."`;
